File: src/main/java/fi/quanfoxes/assembler/builders/Return.java
package fi.quanfoxes.assembler.builders;

import fi.quanfoxes.assembler.*;
import fi.quanfoxes.assembler.references.RegisterReference;
import fi.quanfoxes.parser.nodes.ReturnNode;

public class Return {
    public static Instructions build(Unit unit, ReturnNode node) {
        Instructions instructions = new Instructions();

        Instructions object = References.read(unit, node.first());
        instructions.append(object);

        Reference reference = object.getReference();

        if (reference.getRegister() != unit.eax) {
            instructions.append(Memory.move(unit, reference, new RegisterReference(unit.eax)));
        }

        instructions.append(FunctionBuilder.FOOTER);

        return instructions.setReference(Value.getOperation(unit.eax, reference.getSize()));
    }
}

File: Zigzag/Assembler/Builders/Return.cs
public class Return
{
	public static Instructions Build(Unit unit, ReturnNode node)
	{
		var instructions = new Instructions();

		var @object = References.Read(unit, node.First);
		instructions.Append(@object);

		var reference = @object.Reference;

		if (reference.GetRegister() != unit.EAX)
		{
			Memory.Move(unit, instructions, reference, new RegisterReference(unit.EAX));
		}

		var inline = node.FindParent(n => n.GetNodeType() == NodeType.INLINE_NODE) as InlineNode;

		if (inline == null)
		{
			instructions.Append(Functions.FOOTER);
		}
		else
		{
			instructions.Append($"jmp {inline.GetEndLabel()}");
		}

		return instructions.SetReference(Value.GetOperation(unit.EAX, reference.GetSize()));
	}
}